collect_app/src/main/java/org/odk/collect/android/tasks/sms/SmsSubmissionManagerImpl.java

```diff
@@ -0,0 +1,170 @@
+
+import java.lang.reflect.Type;
+import java.util.Date;
+import java.util.List;
+
+public class SmsSubmissionManagerImpl implements SmsSubmissionManagerContract {
+    private static SharedPreferences preferences;
+    private final SharedPreferences.Editor editor;
+
+    public static final String PREF_FILE_NAME = "submissions_preferences";
+    public static final String KEY_SUBMISSION = "submissions_list_key";
+
+    public SmsSubmissionManagerImpl(Context context) {
+        preferences = context.getSharedPreferences(PREF_FILE_NAME, Context.MODE_PRIVATE);
+        editor = preferences.edit();
+    }
+
+    public SmsSubmission getSubmissionModel(String instanceId) {
+        return getSubmissionFromPrefs(instanceId);
+    }
+
+    @Override
+    public boolean markMessageAsSent(String instanceId, int messageId) {
+
+        SmsSubmission model = getSubmissionModel(instanceId);
+
+        if (model == null) {
+            return false;
+        }
+
+        List<Message> list = model.getMessages();
+
+        boolean updated = false;
+        for (Message message : list) {
+            if (message.getId() == messageId) {
+                message.setMessageStatus(MessageStatus.Sent);
+                list.set(list.indexOf(message), message);
+
+                updated = true;
+            }
+        }
+
+        model.setMessages(list);
+        saveSubmission(model);
+
+        return updated;
+    }
+
+    @Override
+    public boolean markMessageAsSending(String instanceId, int messageId) {
+        SmsSubmission model = getSubmissionModel(instanceId);
+
+        if (model == null) {
+            return false;
+        }
+
+        List<Message> list = model.getMessages();
+
+        boolean updated = false;
+        for (Message message : list) {
+            if (message.getId() == messageId) {
+                message.setMessageStatus(MessageStatus.Sending);
+                list.set(list.indexOf(message), message);
+
+                updated = true;
+            }
+        }
+
+        model.setMessages(list);
+        saveSubmission(model);
+
+        return updated;
+    }
+
+    @Override
+    public boolean updateMessageStatus(MessageStatus messageStatus, String instanceId, int messageId) {
+        SmsSubmission model = getSubmissionModel(instanceId);
+
+        if (model == null) {
+            return false;
+        }
+
+        List<Message> list = model.getMessages();
+
+        boolean updated = false;
+        for (Message message : list) {
+            if (message.getId() == messageId) {
+                message.setMessageStatus(messageStatus);
+                list.set(list.indexOf(message), message);
+
+                updated = true;
+            }
+        }
+
+        model.setMessages(list);
+        model.setLastUpdated(new Date());
+        saveSubmission(model);
+
+        return updated;
+    }
+
+    @Override
+    public void deleteSubmission(String instanceId) {
```

Comment:
I called it `deleteSubmission` because it is conforming to the interface and in this case, the submission is being removed from the prefs. I could rename it though.